plugin/metadata/provider.go

```diff
@@ -1,3 +1,33 @@
+// Package metadata provides an API that allows plugins to add metadata to the context.
+// Each metadata is stored under a label that has the form <plugin>/<name>. Each metadata
+// is returned as a Func. When Func is called the metadata is returned. If Func is expensive to
+// execute it is its responsibility to provide some form of caching. During the handling of a
+// query it is expected the metadata stays constant.
+//
+// Basic example:
+//
+// Implement the Provder interface for a plugin:
+//
+//    func (p P) Metadata(ctx context.Context, state request.Request) context.Context {
+//       cached := ""
+//       f := func() string {
+//		if cached != "" {
+//                 return cached
+//             }
+//             cached = expensiveFunc()
+//             return cached
+//       }
+//       metadata.SetValueFunc(ctx, "test/something", f)
+//	 return ctx
+//    }
```

Comment:
YES!. the inner function is a good utility that could be provided as utility for any metadater. 
Better maybe : could the cache be in the map of the context ?
=> I mean SetValueFunc(ctx, name, fetchingFunction)

Reply:
There is maybe pro and cons .. 
But I am thinking that then we would have one place only to manage the cache and maybe access to this value (race conditition ?)

Reply:
WARNING: I guess we need to support the value "" case => which means maybe have a boolean to remember the cache is set or not.